Problem generating `SDecide` instances for custom promoted types

I followed the [README](https://github.com/goldfirere/singletons#arrows-symbol-and-literals) to get around the `Symbol`/`Text` problem for some types in an application. Now I'm trying to generate `SDecide` instances for my custom promoted type, but it seems the instances are generated incorrectly, i.e., the name mapping for the promoted types is not applied correctly.

I have a minimal example to demonstrate the problem here:

```haskell
{-# LANGUAGE DataKinds                #-}
{-# LANGUAGE FlexibleContexts         #-}
{-# LANGUAGE GADTs                    #-}
{-# LANGUAGE KindSignatures           #-}
{-# LANGUAGE OverloadedStrings        #-}
{-# LANGUAGE QuasiQuotes              #-}
{-# LANGUAGE ScopedTypeVariables      #-}
{-# LANGUAGE StandaloneKindSignatures #-}
{-# LANGUAGE TemplateHaskell          #-}
{-# LANGUAGE TypeApplications         #-}
{-# LANGUAGE TypeFamilies             #-}
{-# LANGUAGE UndecidableInstances     #-}
module Main (main) where

import           Data.Singletons.TH         (genSingletons, singDecideInstances,
                                             singletons)
import           Data.Singletons.TH.Options (defaultOptions,
                                             defunctionalizedName,
                                             promotedDataTypeOrConName,
                                             withOptions)
import           Data.String                (fromString)
import           Data.String.Singletons     (FromString)
import           Data.Text                  (Text)
import           GHC.TypeLits.Singletons    (Symbol)
import           Language.Haskell.TH        (Name)

-- Term-level
newtype Message = MkMessage Text
-- Type-level
newtype PMessage = PMkMessage Symbol

$(let customPromote :: Name -> Name
      customPromote n
        | n == ''Message  = ''PMessage
        | n == 'MkMessage = 'PMkMessage
        | n == ''Text     = ''Symbol
        | otherwise       = promotedDataTypeOrConName defaultOptions n

      customDefun :: Name -> Int -> Name
      customDefun n sat = defunctionalizedName defaultOptions (customPromote n) sat in

  withOptions defaultOptions{ promotedDataTypeOrConName = customPromote
                            , defunctionalizedName      = customDefun
                            } $ do
    decs1 <- genSingletons [''Message]
    decs2 <- singletons [d|
               hello :: Message
               hello = MkMessage "hello"
               |]
    decs3 <- singDecideInstances [''Message]
    return $ decs1 ++ decs2 ++ decs3)

main :: IO ()
main = putStrLn "Hello, Haskell!"
```

This yields the following errors:
```
$ cabal build
Resolving dependencies...
Build profile: -w ghc-9.2.4 -O1
In order, the following will be built (use -v for more details):
 - singletons-decide-instances-0.1.0.0 (exe:singletons-decide-instances) (file app/Main.hs changed)
Preprocessing executable 'singletons-decide-instances' for singletons-decide-instances-0.1.0.0..
Building executable 'singletons-decide-instances' for singletons-decide-instances-0.1.0.0..
[1 of 1] Compiling Main             ( app/Main.hs, /home/bgohla/src/scratch/singletons-decide-instances/dist-newstyle/build/x86_64-openbsd/ghc-9.2.4/singletons-decide-instances-0.1.0.0/x/singletons-decide-instances/build/singletons-decide-instances/singletons-decide-instances-tmp/Main.o, /home/bgohla/src/scratch/singletons-decide-instances/dist-newstyle/build/x86_64-openbsd/ghc-9.2.4/singletons-decide-instances-0.1.0.0/x/singletons-decide-instances/build/singletons-decide-instances/singletons-decide-instances-tmp/Main.dyn_o )

app/Main.hs:32:2: error:
    • Couldn't match kind ‘PMessage’ with ‘Message’
      Expected kind ‘Message -> *’,
        but ‘SMessage’ has kind ‘PMessage -> *’
    • In the first argument of ‘Data.Type.Coercion.TestCoercion’, namely
        ‘(SMessage :: Message -> ghc-prim-0.8.0:GHC.Types.Type)’
      In the instance declaration for
        ‘Data.Type.Coercion.TestCoercion (SMessage :: Message
                                                      -> ghc-prim-0.8.0:GHC.Types.Type)’
   |
32 | $(let customPromote :: Name -> Name
   |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...

app/Main.hs:32:2: error:
    • Couldn't match kind ‘PMessage’ with ‘Message’
      Expected kind ‘Message -> *’,
        but ‘SMessage’ has kind ‘PMessage -> *’
    • In the first argument of ‘Data.Type.Equality.TestEquality’, namely
        ‘(SMessage :: Message -> ghc-prim-0.8.0:GHC.Types.Type)’
      In the instance declaration for
        ‘Data.Type.Equality.TestEquality (SMessage :: Message
                                                      -> ghc-prim-0.8.0:GHC.Types.Type)’
   |
32 | $(let customPromote :: Name -> Name
   |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

So it seems `singDecideInstances` disregards the custom name mapping we defined. I have not dug into the implementation to understand what goes wrong.

A Cabal project with a freeze file is here https://github.com/cgohla/singletons-decide-instances.

--- a/singletons-base/CHANGES.md
+++ b/singletons-base/CHANGES.md
@@ -1,8 +1,8 @@
 Changelog for the `singletons-base` project
 ===========================================
 
-next [????.??.??]
------------------
+3.1.2 [????.??.??]
+------------------
 * Provide `TestEquality` and `TestCoercion` instances for `SNat, `SSymbol`, and
   `SChar`.
 

--- a/singletons-base/singletons-base.cabal
+++ b/singletons-base/singletons-base.cabal
@@ -1,5 +1,5 @@
 name:           singletons-base
-version:        3.1.1
+version:        3.1.2
 cabal-version:  1.24
 synopsis:       A promoted and singled version of the base library
 homepage:       http://www.github.com/goldfirere/singletons
@@ -55,7 +55,7 @@ source-repository this
   type:     git
   location: https://github.com/goldfirere/singletons.git
   subdir:   singletons-base
-  tag:      v3.1.1
+  tag:      v3.1.2
 
 source-repository head
   type:     git
@@ -75,7 +75,7 @@ library
   build-depends:      base             >= 4.17 && < 4.18,
                       pretty,
                       singletons       == 3.0.*,
-                      singletons-th    == 3.1.*,
+                      singletons-th    >= 3.1.2 && < 3.2,
                       template-haskell >= 2.19 && < 2.20,
                       text >= 1.2,
                       th-desugar       >= 1.14 && < 1.15

--- a/singletons-base/tests/SingletonsBaseTestSuite.hs
+++ b/singletons-base/tests/SingletonsBaseTestSuite.hs
@@ -144,6 +144,7 @@ tests =
     , compileAndDumpStdTest "T492"
     , compileAndDumpStdTest "Natural"
     , compileAndDumpStdTest "T511"
+    , compileAndDumpStdTest "T536"
     ],
     testCompileAndDumpGroup "Promote"
     [ compileAndDumpStdTest "Constructors"

--- a/singletons-base/tests/compile-and-dump/Singletons/T536.golden
+++ b/singletons-base/tests/compile-and-dump/Singletons/T536.golden
@@ -1,0 +1,97 @@
+Singletons/T536.hs:(0,0)-(0,0): Splicing declarations
+    let
+      customPromote :: Name -> Name
+      customPromote n
+        | n == ''Message = ''PMessage
+        | n == 'MkMessage = 'PMkMessage
+        | n == ''Text = ''Symbol
+        | otherwise = promotedDataTypeOrConName defaultOptions n
+      customDefun :: Name -> Int -> Name
+      customDefun n sat
+        = defunctionalizedName defaultOptions (customPromote n) sat
+    in
+      withOptions
+        defaultOptions
+          {promotedDataTypeOrConName = customPromote,
+           defunctionalizedName = customDefun}
+        $ do decs1 <- genSingletons [''Message]
+             decs2 <- singletons
+                        [d| hello :: Message
+                            hello = MkMessage "hello" |]
+             decs3 <- singDecideInstances [''Message]
+             decs4 <- showSingInstances [''Message]
+             return $ decs1 ++ decs2 ++ decs3 ++ decs4
+  ======>
+    type PMkMessageSym0 :: (Data.Singletons.~>) Symbol PMessage
+    data PMkMessageSym0 :: (Data.Singletons.~>) Symbol PMessage
+      where
+        PMkMessageSym0KindInference :: Data.Singletons.SameKind (Data.Singletons.Apply PMkMessageSym0 arg) (PMkMessageSym1 arg) =>
+                                       PMkMessageSym0 a0123456789876543210
+    type instance Data.Singletons.Apply PMkMessageSym0 a0123456789876543210 = 'PMkMessage a0123456789876543210
+    instance Data.Singletons.TH.SuppressUnusedWarnings.SuppressUnusedWarnings PMkMessageSym0 where
+      Data.Singletons.TH.SuppressUnusedWarnings.suppressUnusedWarnings
+        = snd (((,) PMkMessageSym0KindInference) ())
+    type PMkMessageSym1 :: Symbol -> PMessage
+    type family PMkMessageSym1 (a0123456789876543210 :: Symbol) :: PMessage where
+      PMkMessageSym1 a0123456789876543210 = 'PMkMessage a0123456789876543210
+    type SMessage :: PMessage -> Type
+    data SMessage :: PMessage -> Type
+      where
+        SMkMessage :: forall (n :: Symbol).
+                      (Data.Singletons.Sing n) -> SMessage ('PMkMessage n :: PMessage)
+    type instance Data.Singletons.Sing @PMessage = SMessage
+    instance Data.Singletons.SingKind PMessage where
+      type Data.Singletons.Demote PMessage = Message
+      Data.Singletons.fromSing (SMkMessage b)
+        = MkMessage (Data.Singletons.fromSing b)
+      Data.Singletons.toSing
+        (MkMessage (b :: Data.Singletons.Demote Symbol))
+        = case  Data.Singletons.toSing b :: Data.Singletons.SomeSing Symbol
+          of
+            Data.Singletons.SomeSing c
+              -> Data.Singletons.SomeSing (SMkMessage c)
+    instance Data.Singletons.SingI n =>
+             Data.Singletons.SingI ('PMkMessage (n :: Symbol)) where
+      Data.Singletons.sing = SMkMessage Data.Singletons.sing
+    instance Data.Singletons.SingI1 'PMkMessage where
+      Data.Singletons.liftSing = SMkMessage
+    instance Data.Singletons.SingI (PMkMessageSym0 :: (Data.Singletons.~>) Symbol PMessage) where
+      Data.Singletons.sing
+        = (Data.Singletons.singFun1 @PMkMessageSym0) SMkMessage
+    hello :: Message
+    hello = MkMessage "hello"
+    type HelloSym0 :: PMessage
+    type family HelloSym0 :: PMessage where
+      HelloSym0 = Hello
+    type Hello :: PMessage
+    type family Hello :: PMessage where
+      Hello = Data.Singletons.Apply PMkMessageSym0 (FromString "hello")
+    sHello :: Data.Singletons.Sing (HelloSym0 :: PMessage)
+    sHello
+      = (Data.Singletons.applySing
+           ((Data.Singletons.singFun1 @PMkMessageSym0) SMkMessage))
+          (sFromString
+             (Data.Singletons.sing :: Data.Singletons.Sing "hello"))
+    instance Data.Singletons.Decide.SDecide Text =>
+             Data.Singletons.Decide.SDecide PMessage where
+      (Data.Singletons.Decide.%~) (SMkMessage a) (SMkMessage b)
+        = case ((Data.Singletons.Decide.%~) a) b of
+            Data.Singletons.Decide.Proved Data.Type.Equality.Refl
+              -> Data.Singletons.Decide.Proved Data.Type.Equality.Refl
+            Data.Singletons.Decide.Disproved contra
+              -> Data.Singletons.Decide.Disproved
+                   (\ refl
+                      -> case refl of
+                           Data.Type.Equality.Refl -> contra Data.Type.Equality.Refl)
+    instance Data.Singletons.Decide.SDecide Text =>
+             Data.Type.Equality.TestEquality (SMessage :: PMessage
+                                                          -> Type) where
+      Data.Type.Equality.testEquality
+        = Data.Singletons.Decide.decideEquality
+    instance Data.Singletons.Decide.SDecide Text =>
+             Data.Type.Coercion.TestCoercion (SMessage :: PMessage
+                                                          -> Type) where
+      Data.Type.Coercion.testCoercion
+        = Data.Singletons.Decide.decideCoercion
+    deriving instance Data.Singletons.ShowSing.ShowSing Text =>
+                      Show (SMessage (z :: PMessage))

--- a/singletons-base/tests/compile-and-dump/Singletons/T536.hs
+++ b/singletons-base/tests/compile-and-dump/Singletons/T536.hs
@@ -1,0 +1,41 @@
+{-# LANGUAGE OverloadedStrings #-}
+module T536 where
+
+import           Data.Singletons.TH         (genSingletons, showSingInstances,
+                                             singDecideInstances, singletons)
+import           Data.Singletons.TH.Options (defaultOptions,
+                                             defunctionalizedName,
+                                             promotedDataTypeOrConName,
+                                             withOptions)
+import           Data.String                (fromString)
+import           Data.String.Singletons     (FromString, sFromString)
+import           Data.Text                  (Text)
+import           GHC.TypeLits.Singletons    (Symbol)
+import           Language.Haskell.TH        (Name)
+
+-- Term-level
+newtype Message = MkMessage Text
+-- Type-level
+newtype PMessage = PMkMessage Symbol
+
+$(let customPromote :: Name -> Name
+      customPromote n
+        | n == ''Message  = ''PMessage
+        | n == 'MkMessage = 'PMkMessage
+        | n == ''Text     = ''Symbol
+        | otherwise       = promotedDataTypeOrConName defaultOptions n
+
+      customDefun :: Name -> Int -> Name
+      customDefun n sat = defunctionalizedName defaultOptions (customPromote n) sat in
+
+  withOptions defaultOptions{ promotedDataTypeOrConName = customPromote
+                            , defunctionalizedName      = customDefun
+                            } $ do
+    decs1 <- genSingletons [''Message]
+    decs2 <- singletons [d|
+               hello :: Message
+               hello = MkMessage "hello"
+               |]
+    decs3 <- singDecideInstances [''Message]
+    decs4 <- showSingInstances [''Message]
+    return $ decs1 ++ decs2 ++ decs3 ++ decs4)

--- a/singletons-th/CHANGES.md
+++ b/singletons-th/CHANGES.md
@@ -1,6 +1,12 @@
 Changelog for the `singletons-th` project
 =========================================
 
+3.1.2 [????.??.??]
+------------------
+* Fix a bug in which the `singDecideInstances` and `showSingInstances`, as well
+  as `deriving Show` declarations, would not respect custom
+  `promotedDataTypeOrConName` options.
+
 3.1.1 [2022.08.23]
 ------------------
 * Require building with GHC 9.4.

--- a/singletons-th/singletons-th.cabal
+++ b/singletons-th/singletons-th.cabal
@@ -1,5 +1,5 @@
 name:           singletons-th
-version:        3.1.1
+version:        3.1.2
 cabal-version:  1.24
 synopsis:       A framework for generating singleton types
 homepage:       http://www.github.com/goldfirere/singletons
@@ -42,7 +42,7 @@ source-repository this
   type:     git
   location: https://github.com/goldfirere/singletons.git
   subdir:   singletons-th
-  tag:      v3.1.1
+  tag:      v3.1.2
 
 source-repository head
   type:     git

--- a/singletons-th/src/Data/Singletons/TH/Single.hs
+++ b/singletons-th/src/Data/Singletons/TH/Single.hs
@@ -151,11 +151,9 @@ singDecideInstance name = do
   dtvbs <- mapM dsTvbUnit tvbs
   let data_ty = foldTypeTvbs (DConT name) dtvbs
   dcons <- concatMapM (dsCon dtvbs data_ty) cons
-  let tyvars = map (DVarT . extractTvbName) dtvbs
-      kind = foldType (DConT name) tyvars
   (scons, _) <- singM [] $ mapM (singCtor name) dcons
-  sDecideInstance <- mkDecideInstance Nothing kind dcons scons
-  testInstances <- traverse (mkTestInstance Nothing kind name dcons)
+  sDecideInstance <- mkDecideInstance Nothing data_ty dcons scons
+  testInstances <- traverse (mkTestInstance Nothing data_ty name dcons)
                             [TestEquality, TestCoercion]
   return $ decsToTH (sDecideInstance:testInstances)
 
@@ -1023,9 +1021,10 @@ singDerivedShowDecs (DerivedDecl { ded_mb_cxt     = mb_cxt
     show_cxt <- inferConstraintsDef (fmap mkShowSingContext mb_cxt)
                                     (DConT showSingName)
                                     ty cons
+    ki <- promoteType ty
     let sty_tycon = singledDataTypeName opts ty_tycon
         show_inst = DStandaloneDerivD Nothing Nothing show_cxt
-                      (DConT showName `DAppT` (DConT sty_tycon `DAppT` DSigT (DVarT z) ty))
+                      (DConT showName `DAppT` (DConT sty_tycon `DAppT` DSigT (DVarT z) ki))
     pure [show_inst]
 
 isException :: DExp -> Bool

--- a/singletons-th/src/Data/Singletons/TH/Single/Decide.hs
+++ b/singletons-th/src/Data/Singletons/TH/Single/Decide.hs
@@ -14,41 +14,44 @@ import Language.Haskell.TH.Desugar
 import Data.Singletons.TH.Deriving.Infer
 import Data.Singletons.TH.Names
 import Data.Singletons.TH.Options
+import Data.Singletons.TH.Promote.Type
 import Data.Singletons.TH.Util
 import Control.Monad
 
 -- Make an instance of SDecide.
-mkDecideInstance :: DsMonad q => Maybe DCxt -> DKind
+mkDecideInstance :: OptionsMonad q => Maybe DCxt -> DType
                  -> [DCon] -- ^ The /original/ constructors (for inferring the instance context)
                  -> [DCon] -- ^ The /singletons/ constructors
                  -> q DDec
-mkDecideInstance mb_ctxt k ctors sctors = do
+mkDecideInstance mb_ctxt data_ty ctors sctors = do
   let sctorPairs = [ (sc1, sc2) | sc1 <- sctors, sc2 <- sctors ]
   methClauses <- if null sctors
                  then (:[]) <$> mkEmptyDecideMethClause
                  else mapM mkDecideMethClause sctorPairs
-  constraints <- inferConstraintsDef mb_ctxt (DConT sDecideClassName) k ctors
+  constraints <- inferConstraintsDef mb_ctxt (DConT sDecideClassName) data_ty ctors
+  data_ki <- promoteType data_ty
   return $ DInstanceD Nothing Nothing
                      constraints
-                     (DAppT (DConT sDecideClassName) k)
+                     (DAppT (DConT sDecideClassName) data_ki)
                      [DLetDec $ DFunD sDecideMethName methClauses]
 
 data TestInstance = TestEquality
                   | TestCoercion
 
 -- Make an instance of TestEquality or TestCoercion by leveraging SDecide.
-mkTestInstance :: OptionsMonad q => Maybe DCxt -> DKind
+mkTestInstance :: OptionsMonad q => Maybe DCxt -> DType
                -> Name   -- ^ The name of the data type
                -> [DCon] -- ^ The /original/ constructors (for inferring the instance context)
                -> TestInstance -> q DDec
-mkTestInstance mb_ctxt k data_name ctors ti = do
+mkTestInstance mb_ctxt data_ty data_name ctors ti = do
   opts <- getOptions
-  constraints <- inferConstraintsDef mb_ctxt (DConT sDecideClassName) k ctors
+  constraints <- inferConstraintsDef mb_ctxt (DConT sDecideClassName) data_ty ctors
+  data_ki <- promoteType data_ty
   pure $ DInstanceD Nothing Nothing
                     constraints
                     (DAppT (DConT tiClassName)
                            (DConT (singledDataTypeName opts data_name)
-                             `DSigT` (DArrowT `DAppT` k `DAppT` DConT typeKindName)))
+                             `DSigT` (DArrowT `DAppT` data_ki `DAppT` DConT typeKindName)))
                     [DLetDec $ DFunD tiMethName
                                      [DClause [] (DVarE tiDefaultName)]]
   where

